### PR TITLE
✨ feat: add work version registry

### PR DESCRIPTION
**Pull request by tjx666**
#### 💻 Change Type

- [x] ✨ feat
- [x] 🐛 fix
- [x] ♻️ refactor
- [ ] 💄 style
- [ ] 👷 build
- [ ] ⚡️ perf
- [x] ✅ test
- [ ] 📝 docs
- [x] 🔨 chore

#### 🔗 Related Issue

Related to LOBE-10820 (Work MVP), LOBE-10964 (task adaptation), LOBE-10966 (Linear adaptation), LOBE-10967 (GitHub adaptation)

#### 🔀 Description of Change

Add the initial Work version registry for task resources and attribution snapshots:

- add `works` and `work_versions` database tables with migration metadata — a version row is the mutation event itself (git-commit mental model), carrying both the resource snapshot and its attribution (`role`, `source`, `topicId`/`threadId`, `sourceMessageId`, `sourceToolCallId`, `rootOperationId`, `actorAgentId`, cost/usage)
- register task tool results as Work versions, including task creation, updates, and verify configuration changes
- expose Work lambda APIs for conversation/sidebar queries and assistant-message display anchors
- render task Work versions in assistant messages and the working sidebar
- capture cumulative operation cost in USD on each Work version and store usage/cost details in JSONB
- display Work version cumulative cost in assistant Work cards and the working sidebar version list
- attach client-runtime tool Work records to the final assistant message for stable two-round display
- make Work migration reruns idempotent for preview deployments
- pin `@lobehub/editor` to `4.18.0` to avoid the `4.19.0` `FloatMenu` export regression in no-lockfile CI installs

Skill provider adaptations (LOBE-10966 / LOBE-10967):

- generalize the Work registration entry to a provider-dispatched `handleSkillToolResult` (server + client hooks), gated by `WORK_SKILL_PROVIDERS`
- register Linear issue/document create & update tool results as Work versions with patch-field snapshot merging
- register GitHub issue/PR create & update tool results as Work versions, covering both skill surfaces: structured REST-shaped tools (`create_issue` / `update_issue` / `create_pull_request` / `update_pull_request`) and the sandbox `runCommand` surface that executes `gh` CLI (the dominant one in production) by parsing `gh issue/pr create|edit` commands plus the stdout entity URL
- github Work identity is `owner/repo#number` (both `resourceId` and `resourceIdentifier`): the gh CLI surface never returns a `node_id`, so both surfaces must share this dedup key to land on one Work row
- read-only queries, failed results (isError / non-zero exitCode), and unidentifiable results are never registered
- render Linear/GitHub Work summaries and version history in the conversation and working sidebar (opens the external URL)

Schema simplification (two-table design):

- drop the `work_contexts` table: it was strictly 1:1 with `work_versions` in every write path, so the attribution fields are flattened onto the version row and list/summary queries read a single table (summaries self-join `work_versions` via alias to pair the filtered event row with the Work's current version)
- drop both `title` columns (`works.title`, `work_versions.title`): display names come from the resource snapshot (task name is read live from the `tasks` join); there is deliberately **no synthesized fallback title** — a nameless resource shows only its identifier so data gaps stay visible
- drop `source_type` (only `'tool'` was ever written) and the snapshot `entityType` field (derivable from `works.resourceType`)
- narrow version `role` to `'created' | 'updated'`; it stays a stored column because it is not derivable from `version === 1` (updating a never-registered external resource yields a v1 row with `role='updated'`)

#### 🧭 Key Decisions / Non-goals

- Work version cost is a cumulative operation snapshot, not exclusive per-Work cost allocation.
- Only `cumulative_cost` is promoted as a top-level USD numeric column; token/cost breakdown details stay in JSONB.
- Work cards in the sidebar start collapsed so opening the working panel does not request versions for every work row immediately.
- Linear Work support covers issue and document entities; the earlier comment adaptation was dropped from this PR.
- Linear/GitHub snapshots persist only normalized Work fields. Raw connector responses are intentionally not stored.
- The two-table flattening regenerates the unreleased `0116` migration in place (no compatibility migration) since the feature has not shipped.

#### 🧪 How to Test

- [x] Tested locally
- [x] Added/updated tests
- [ ] No tests needed

Automated tests:

```bash
cd packages/database && bunx vitest run src/models/__tests__/work.test.ts
bunx vitest run apps/server/src/services/toolExecution/serverRuntimes/__tests__/task.test.ts
bunx vitest run apps/server/src/modules/AgentRuntime/__tests__/dispatchClientTool.test.ts
bunx vitest run apps/server/src/services/toolExecution/__tests__/builtin.test.ts
```

Manual verification:

```bash
DATABASE_DRIVER=node bun run db:migrate
```

#### 📸 Screenshots / Videos

N/A

#### 📝 Additional Information

This PR includes the `0116_add_work_registry_tables` database migration. The cloud gitlink PR (lobehub-biz/lobehub-cloud#916) tracks this branch.

### CODE REVIEW

File: packages/database/migrations/0116_add_work_registry_tables.sql [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"resource_type" text NOT NULL,
+	"resource_id" text NOT NULL,
+	"resource_identifier" text,
+	"user_id" text NOT NULL,
+	"workspace_id" text,
+	"created_at" timestamp with time zone DEFAULT now() NOT NULL,
+	"updated_at" timestamp with time zone DEFAULT now() NOT NULL
+);
+--> statement-breakpoint
+ALTER TABLE "work_contexts" ADD CONSTRAINT "work_contexts_work_id_works_id_fk" FOREIGN KEY ("work_id") REFERENCES "public"."works"("id") ON DELETE cascade ON UPDATE no action;--> statement-breakpoint
```
[the remark this answers is not in the document]
Comment (tjx666) [reply packages/database/migrations/0116_add_work_registry_tables.sql]: 已在当前 HEAD 处理：0115 migration 的每个 ADD CONSTRAINT 前都补了 DROP CONSTRAINT IF EXISTS，满足可重入执行要求。

File: src/routes/(main)/agent/features/Conversation/WorkingSidebar/WorksSection/index.tsx [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        );
+      })}
+    </Flexbox>
+  );
+});
+
+VersionList.displayName = 'VersionList';
+
+const WorkCard = memo<{ work: TaskWorkListItem }>(({ work }) => {
+  const [expanded, setExpanded] = useState(true);
```
[the remark this answers is not in the document]
Comment (tjx666) [reply src/routes/(main)/agent/features/Conversation/WorkingSidebar/WorksSection/index.tsx]: 已处理：Work 卡片默认改为折叠，只有用户展开时才挂载 VersionList 并请求版本列表，避免打开工作面板时为所有 work 立即发起 listVersions。

**Comment by tjx666** [comment]
@codex review latest code

**Comment by tjx666** [comment]
@codex review latest code

**Comment by tjx666** [comment]
@codex fix the following review findings on the Linear → Work registry adaptation (commits f42f4737 + 5c3b871b):

**1. Major — partial update responses erase fields from the current snapshot**
`createRegisterOperation` (`packages/database/src/models/work/linearToolResult.ts`) builds the snapshot solely from the current tool result, and `linearSnapshot` (`packages/database/src/models/work.ts`) never merges with the previous version. Linear update responses are routinely partial (the test fixtures model this: the second `save_issue` omits `team`/`priority`/`labels`). Since every summary/sidebar read path joins on `works.currentVersionId`, a state-only update blanks out priority/team/labels/assignee in the UI even though they are still set on the real Linear resource. Fix by merging the new snapshot over the previous version's snapshot (field-level, only overwrite when the new result actually provides a value), and add a regression test asserting the latest version retains fields omitted by a partial update.

**2. Major — `resourceIdentifier` "preserve on missing" never engages for issue/document**
The preserve logic relies on `COALESCE(new, old)` in `upsertLinearWork`, which only helps when the new value is `null`. But `resolveResourceIdentifier` ends its fallback chain with `?? params.title ?? params.id` for both `issue` and `document`, so the computed identifier is never null — a partial update missing slug/url/identifier overwrites the human-readable identifier with the raw internal id instead of preserving it. `works.title` has the same problem (unconditional overwrite via `title: title ?? identifier ?? id`). Make the fallback chain stop before degrading to `params.id`/`params.title` so COALESCE can preserve the existing value, and apply the same preserve-on-missing behavior to `title`.

**3. Major (test gap) — missing cross-user isolation test for Linear works**
`work.test.ts` has "does not let another user register someone else …" tests for document and task works, but none for `handleLinearToolResult`/`deleteLinearWork`. Add the analogous two-user test covering register, update, and delete.

**4. Minor — read-only Linear tools trigger a pointless mutation + SWR refresh**
Client path (`src/store/chat/slices/plugin/actions/exector.ts`) gates only on `provider === 'linear'`, so every successful `get_issue`/`list_issues` call round-trips `work.handleLinearToolResult` (which returns null) and then `workService.handleLinearToolResult` unconditionally awaits `refreshConversation` + `refreshRootOperation`. Only refresh when a work item was actually returned.

**5. Minor — `update_document` is a dead switch case**
`LINEAR_TOOL_NAMES` (`packages/shared-tool-ui/src/Inspector/Linear/labels.ts`) has `create_document`/`save_document` but no `update_document`. Remove the dead case (or correct the tool name if one is actually missing).

**6. Minor — `snapshotText` truncation can split a surrogate pair**
`text.slice(0, MAX_LINEAR_SNAPSHOT_TEXT_LENGTH)` operates on UTF-16 code units; an emoji straddling the 4000 boundary becomes a lone surrogate. Trim the trailing lone surrogate (or use a code-point-safe slice).

**7. Minor — `builtin.ts` parse fallback uses `??` instead of the `=== undefined` idiom**
`safeParseJSON(result.content) ?? result.content` can't distinguish a parsed `null` from a parse failure; the same file already uses `parsed === undefined` two hunks above. Align for consistency.

Please keep the fixes scoped to these findings, follow the existing task/document Work patterns, and update/extend the tests in `packages/database/src/models/__tests__/work.test.ts` accordingly.

File: src/routes/(main)/agent/features/Conversation/WorkingSidebar/WorksSection/index.tsx [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+        );
+      })}
+    </Flexbox>
+  );
+});
+
+VersionList.displayName = 'VersionList';
+
+const TaskWorkVersionHistoryCard = memo<{ work: TaskWorkListItem }>(({ work }) => {
+  const [expanded, setExpanded] = useState(true);
```
[the remark this answers is not in the document]
Comment (tjx666) [reply src/routes/(main)/agent/features/Conversation/WorkingSidebar/WorksSection/index.tsx]: Fixed in 71cd8cb9b1 — all four history cards (task/document/linear/github) now default to collapsed, so switching to History no longer fans out listVersions requests; versions load only after the user expands a row.

File: packages/database/src/models/work/linearToolResult.ts [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+const fromRecord = (record: Record<string, unknown>, keys: string[]) => {
+  for (const key of keys) {
+    const value = stringValue(record[key]);
+    if (value) return value;
+  }
+
+  return null;
+};
+
+const textFromRecord = (record: Record<string, unknown>, keys: string[]) => {
```
[the remark this answers is not in the document]
Comment (tjx666) [reply packages/database/src/models/work/linearToolResult.ts]: Fixed in 71cd8cb9b1 — removed the unused textFromRecord helper.

File: packages/database/src/models/work/linearToolResult.ts [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+  for (const key of keys) {
+    const value = snapshotText(record[key]);
+    if (value) return value;
+  }
+
+  return null;
+};
+
+const fromNestedRecord = (record: Record<string, unknown>, key: string, keys: string[]) =>
```
[the remark this answers is not in the document]
Comment (tjx666) [reply packages/database/src/models/work/linearToolResult.ts]: Fixed in 71cd8cb9b1 — removed the unused fromNestedRecord helper.

File: packages/database/src/models/work/linearToolResult.ts [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+  fromRecord(toRecord(record[key]) ?? {}, keys);
+
+const numberFromRecord = (record: Record<string, unknown>, keys: string[]) => {
```
[the remark this answers is not in the document]
Comment (tjx666) [reply packages/database/src/models/work/linearToolResult.ts]: Fixed in 71cd8cb9b1 — removed the unused numberFromRecord helper.

**Comment by tjx666** [comment]
@codex review latest code

**Comment by tjx666** [comment]
@codex review latest code

File: packages/database/src/models/work/writes.ts [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  ctx: WorkContext,
+  work: WorkItem,
+  params: WorkVersionEventParams,
+  buildInput: () => CreateVersionInput | Promise<CreateVersionInput>,
+): Promise<WorkVersionItem> => {
+  const existing = await findVersionBySourceToolCall(ctx, work.id, params.sourceToolCallId);
+  if (existing) return existing;
+
+  for (let attempt = 0; attempt < MAX_VERSION_CREATE_RETRIES; attempt += 1) {
+    const { metadata, snapshot } = await buildInput();
```
[the remark this answers is not in the document]
Comment (tjx666) [reply packages/database/src/models/work/writes.ts]: Fixed in b68ce80777: `buildInput` now runs inside the transaction after a `SELECT … FOR UPDATE` on the works row (same pattern as `TopicModel.updateMetadata`), so the merge base always reads the concurrent winner's committed snapshot; also re-checks sourceToolCallId dedupe under the lock.

File: src/routes/(main)/resource/(home)/_layout/WorkBody/index.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      <Flexbox gap={1} paddingInline={4}>
+        {WORK_GALLERY_ENTRIES.map((entry) => {
+          const url = `/resource?works=${entry.key}`;
+          return (
+            <Link
+              key={entry.key}
+              to={url}
+              onClick={(e) => {
+                e.preventDefault();
+                navigate(url, { replace: true });
```
[the remark this answers is not in the document]
Comment (tjx666) [reply src/routes/(main)/resource/(home)/_layout/WorkBody/index.tsx:44]: Acknowledged, but keeping as-is for consistency: the sibling `CategoryMenu` in the same sidebar (base-branch code) uses the identical raw-href + workspace-aware onClick pattern, so WorkBody follows the established local precedent. Making hrefs workspace-aware is a sidebar-wide change better done uniformly in a follow-up rather than for one nav group in this PR.

**Comment by tjx666** [comment]
@codex review latest code

File: apps/server/src/routers/lambda/work.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    },
+  });
+});
+
+const workProcedureWrite = workProcedure.use(withScopedPermission('agent:update'));
+
+const versionRoleSchema = z.enum(['created', 'updated']);
+
+const registerTaskSchema = z.object({
+  actorAgentId: z.string().nullable().optional(),
```
[the remark this answers is not in the document]
Comment (tjx666) [reply apps/server/src/routers/lambda/work.ts:41]: Fixed in e3de93c062: added `cumulativeCost` / `cumulativeUsage` to `registerTaskSchema` and `registerSkillToolResultSchema` (matching `registerDocumentSchema`), with a comment noting why every register* schema must declare them.

File: packages/agent-gateway-client/src/types.ts [hunk 10]
```diff
@@ -168,16 +168,38 @@ export interface AgentInterventionResponseData {
...
+  /** Current page document ID for page-scoped conversations. */
+  documentId?: string | null;
   /** Per-invocation deadline. Server caps against its own function budget. */
   executionTimeoutMs: number;
+  /** Group chat ID, when the run belongs to a group conversation. */
+  groupId?: string | null;
   /** Tool plugin identifier (e.g. "local-system"). */
   identifier: string;
+  /** Root server-side runtime operation ID for this assistant run. */
+  rootOperationId?: string;
```
[the remark this answers is not in the document]
Comment (tjx666) [reply packages/agent-gateway-client/src/types.ts:188]: The premise does not hold in the current wiring: `executor = 'client'` is only assigned to local-system / stdio MCP / stdio connector tools, and only in standalone deployments without a device gateway (see the `!gatewayConfigured` block in `apps/server/src/services/aiAgent/index.ts`) — none of those declare a `work` config or emit Work intents. Task and agent-document tools declare no client executors in their manifests, so in gateway mode they always execute server-side, where `resolveBuiltinToolWorkIntent` / the `onWorkRegistration` sink already produce Work rows. The renderer stash serves only the legacy client-first `call_tool` runtime, which drains it. If a work-config tool ever gains a client executor, forwarding the intent over `tool_result` would indeed be needed — but no such tool exists today.

**Comment by tjx666** [comment]
@codex review latest code

File: src/store/tool/slices/builtin/executors/index.ts [hunk 11]
```diff
@@ -120,7 +121,15 @@ export const invokeExecutor = async (
...
   }
 
-  return executor.invoke(apiName, params, ctx);
+  const result = await executor.invoke(apiName, params, ctx);
+
+  // Manifest-driven Work registration (best-effort; a no-op unless the API
+  // declares a `work` config). Only STASH the intent here — `call_tool` drains
+  // it and writes the Work version once the tool call's cumulative cost is known
+  // (write-once instead of register-then-backfill).
+  stashBuiltinToolWorkIntent(identifier, apiName, params, ctx, result);
```
[the remark this answers is not in the document]
Comment (tjx666) [reply src/store/tool/slices/builtin/executors/index.ts:130]: Duplicate of the previous round's finding (see the resolved thread on `packages/agent-gateway-client/src/types.ts`): the premise does not hold in the current wiring. `executor = 'client'` is only ever assigned to local-system / stdio MCP / stdio connector tools, and only in standalone deployments without a device gateway (`!gatewayConfigured` block in `apps/server/src/services/aiAgent/index.ts`). None of those declare a `work` config. Task and agent-document tools declare no client executors, so in gateway mode they always execute server-side, where `resolveBuiltinToolWorkIntent` / the `onWorkRegistration` sink already produce Work rows; this stash serves only the legacy client-first `call_tool` runtime, which drains it. Forwarding intents over `tool_result` becomes necessary only if a work-config tool ever gains a client executor — none exists today.

**Comment by tjx666** [comment]
Superseded by #17079 — same content squashed into a single commit after rebasing onto the latest canary.
